File: packages/telemetry/src/client/prometheus-client.ts
import { PrometheusClientConfigSchema } from '@ff-promo/contracts';
import type { PrometheusClientConfig } from '@ff-promo/contracts';
import pRetry from 'p-retry';
import { TelemetryApiError } from '../errors/telemetry-adapter-error.js';

export const DEFAULT_PROMETHEUS_BASE_URL = 'http://localhost:9090';

type PrometheusQueryEnvelope = {
	status: string;
	data?: {
		resultType: string;
		result: unknown;
	};
	error?: string;
	errorType?: string;
};

export type PrometheusInstantQueryData = {
	resultType: string;
	result: unknown;
};

export type PrometheusClient = {
	config: PrometheusClientConfig & { baseUrl: string };
	queryInstant: (
		query: string,
		opts?: { timeout?: string },
	) => Promise<PrometheusInstantQueryData>;
};

function assertHttpOrHttpsUrl(baseUrl: string): void {
	let parsed: URL;
	try {
		parsed = new URL(baseUrl);
	} catch {
		throw new TelemetryApiError('Invalid Prometheus baseUrl', 0, undefined, {
			baseUrl,
		});
	}
	if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
		throw new TelemetryApiError(
			'Prometheus baseUrl must use http or https',
			0,
			undefined,
			{ baseUrl },
		);
	}
}

async function parseEnvelope(
	res: Response,
): Promise<PrometheusInstantQueryData> {
	let body: PrometheusQueryEnvelope;
	try {
		body = (await res.json()) as PrometheusQueryEnvelope;
	} catch {
		throw new TelemetryApiError(
			`Prometheus query failed (${res.status})`,
			res.status,
		);
	}
	if (!res.ok || body.status !== 'success' || !body.data) {
		throw new TelemetryApiError(
			body.error ?? `Prometheus query failed (${res.status})`,
			res.status,
			body,
			{ errorType: body.errorType },
		);
	}
	return body.data;
}

export function createPrometheusClient(
	configInput: PrometheusClientConfig,
): PrometheusClient {
	const config = PrometheusClientConfigSchema.parse(configInput);
	const baseUrl = (
		config.baseUrl ??
		process.env.PROMETHEUS_BASE_URL ??
		DEFAULT_PROMETHEUS_BASE_URL
	).replace(/\/+$/, '');
	const bearerToken = config.bearerToken ?? process.env.PROMETHEUS_BEARER_TOKEN;

	assertHttpOrHttpsUrl(baseUrl);

	return {
		config: { ...config, baseUrl },
		async queryInstant(query: string, opts?: { timeout?: string }) {
			const url = new URL('/api/v1/query', baseUrl);
			url.searchParams.set('query', query);
			if (opts?.timeout) {
				url.searchParams.set('timeout', opts.timeout);
			}
			if (config.timeout && !opts?.timeout) {
				url.searchParams.set('timeout', config.timeout);
			}

			const headers: Record<string, string> = {};
			if (bearerToken) {
				headers.Authorization = `Bearer ${bearerToken}`;
			}

			return pRetry(
				async () => {
					const res = await fetch(url, { headers });
					return parseEnvelope(res);
				},
				{
					retries: 2,
					shouldRetry: (error) => {
						if (error instanceof TelemetryApiError) {
							return error.status === 503;
						}
						return false;
					},
				},
			);
		},
	};
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/client/PrometheusClient.kt
package com.ffpromo.telemetry.client

import com.ffpromo.contracts.PrometheusClientConfig
import com.ffpromo.telemetry.errors.TelemetryApiError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URI

const val DEFAULT_PROMETHEUS_BASE_URL = "http://localhost:9090"

data class ResolvedPrometheusConfig(
    val baseUrl: String,
    val bearerToken: String? = null,
    val timeout: String? = null,
)

data class PrometheusInstantQueryData(
    val resultType: String,
    val result: JsonElement,
)

interface PrometheusClient {
    val config: ResolvedPrometheusConfig

    suspend fun queryInstant(query: String, timeout: String? = null): PrometheusInstantQueryData
}

@Serializable
private data class PrometheusQueryEnvelope(
    val status: String,
    val data: PrometheusInstantQueryDataDto? = null,
    val error: String? = null,
    @SerialName("errorType")
    val errorType: String? = null,
)

@Serializable
private data class PrometheusInstantQueryDataDto(
    val resultType: String,
    val result: JsonElement,
)

private val json = Json { ignoreUnknownKeys = true }

private fun assertHttpOrHttpsUrl(baseUrl: String) {
    val parsed = try {
        URI(baseUrl)
    } catch (_: Exception) {
        throw TelemetryApiError("Invalid Prometheus baseUrl", 0, null, mapOf("baseUrl" to baseUrl))
    }
    if (parsed.scheme != "http" && parsed.scheme != "https") {
        throw TelemetryApiError(
            "Prometheus baseUrl must use http or https",
            0,
            null,
            mapOf("baseUrl" to baseUrl),
        )
    }
}

private fun parseEnvelope(statusCode: Int, bodyText: String): PrometheusInstantQueryData {
    val body = try {
        json.decodeFromString<PrometheusQueryEnvelope>(bodyText)
    } catch (_: Exception) {
        throw TelemetryApiError("Prometheus query failed ($statusCode)", statusCode)
    }

    if (statusCode !in 200..299 || body.status != "success" || body.data == null) {
        throw TelemetryApiError(
            body.error ?: "Prometheus query failed ($statusCode)",
            statusCode,
            body,
            buildMap {
                body.errorType?.let { put("errorType", it) }
            }.ifEmpty { null },
        )
    }

    return PrometheusInstantQueryData(
        resultType = body.data.resultType,
        result = body.data.result,
    )
}

private class OkHttpPrometheusClient(
    override val config: ResolvedPrometheusConfig,
    private val httpClient: OkHttpClient,
) : PrometheusClient {
    override suspend fun queryInstant(query: String, timeout: String?): PrometheusInstantQueryData =
        withContext(Dispatchers.IO) {
            val urlBuilder = "${config.baseUrl}/api/v1/query".toHttpUrl().newBuilder()
                .addQueryParameter("query", query)

            val effectiveTimeout = timeout ?: config.timeout
            if (effectiveTimeout != null) {
                urlBuilder.addQueryParameter("timeout", effectiveTimeout)
            }

            val requestBuilder = Request.Builder().url(urlBuilder.build())
            config.bearerToken?.let { token ->
                requestBuilder.header("Authorization", "Bearer $token")
            }

            var lastError: TelemetryApiError? = null
            repeat(3) { attempt ->
                httpClient.newCall(requestBuilder.build()).execute().use { response ->
                    val bodyText = response.body?.string().orEmpty()
                    try {
                        return@withContext parseEnvelope(response.code, bodyText)
                    } catch (error: TelemetryApiError) {
                        lastError = error
                        if (error.status == 503 && attempt < 2) {
                            return@use
                        }
                        throw error
                    }
                }
            }
            throw lastError ?: TelemetryApiError("Prometheus query failed", 0)
        }
}

fun createPrometheusClient(configInput: PrometheusClientConfig): PrometheusClient {
    val baseUrl = (
        configInput.baseUrl
            ?: System.getenv("PROMETHEUS_BASE_URL")
            ?: DEFAULT_PROMETHEUS_BASE_URL
        ).trimEnd('/')
    val bearerToken = configInput.bearerToken ?: System.getenv("PROMETHEUS_BEARER_TOKEN")

    assertHttpOrHttpsUrl(baseUrl)

    val resolved = ResolvedPrometheusConfig(
        baseUrl = baseUrl,
        bearerToken = bearerToken,
        timeout = configInput.timeout,
    )

    return OkHttpPrometheusClient(resolved, OkHttpClient())
}
